cpp/independent_set/ExactRecursiveIndependentSet.cpp

```diff
@@ -0,0 +1,755 @@
+        connectedEdges = getConnectedEdges(justV);
+        removeElements(justV);
+        std::vector<NetworKit::node> setWithoutV = recursive();
+        restoreElements(justV, connectedEdges);
+
+        return setWithV.size() > setWithoutV.size() ? setWithV : setWithoutV;
+    }
+
+    NetworKit::node lastIndex = std::numeric_limits<NetworKit::node>::min();
+    graph->forNodes([&](NetworKit::node u) {
+        lastIndex = u;
+    });
+    std::vector<bool> visited(lastIndex + 1, false); 
+    dfs(v, visited);
+
+    std::vector<NetworKit::node> component;
+    std::vector<NetworKit::node> theRest;
+    std::vector<NetworKit::node> allVertices;
+    graph->forNodes([&](NetworKit::node u) {
+        if (visited[u]) {
+            component.push_back(u);
+        }
+        else {
+            theRest.push_back(u);
+        }
+        allVertices.push_back(u);
+    });
+
+    if (component.size() != allVertices.size()) {
+        EdgeSet allEdges = getAllEdges();
+
+        removeElements(theRest);
+        std::vector<NetworKit::node> setComponent = recursive();
+        removeElements(component);
+        restoreElements(allVertices, allEdges);
+
+        removeElements(component);
+        std::vector<NetworKit::node> setTheRest = recursive();
+        removeElements(theRest);
+        restoreElements(allVertices, allEdges);
+
+        setComponent.insert(setComponent.end(), setTheRest.begin(), setTheRest.end());
+        return setComponent;
+    }
+
+    bool hasDegree4 = false;
+    bool hasDegree5 = false;
+    graph->forNodes([&](NetworKit::node u) { // all must be 4 or 5
+        if (graph->degree(u) == 4) {
+            hasDegree4 = true;
+        }
+        else {
+            hasDegree5 = true;
+        }
+    });
+
+    if (!(hasDegree4 && hasDegree5)) {
+        std::vector<NetworKit::node> neighborsPlus = getNeighborsPlus(v);
+        EdgeSet connectedEdges = getConnectedEdges(neighborsPlus);
+        removeElements(neighborsPlus);
+        std::vector<NetworKit::node> setWithV = recursive();
+        setWithV.push_back(v);
+        restoreElements(neighborsPlus, connectedEdges);
+
+        std::vector<NetworKit::node> vWithMirrors = getMirrors(v);
+        vWithMirrors.push_back(v);
+        connectedEdges = getConnectedEdges(vWithMirrors);
+        removeElements(vWithMirrors);
+        std::vector<NetworKit::node> setWithoutV = recursive();
+        restoreElements(vWithMirrors, connectedEdges);
+
+        return setWithV.size() > setWithoutV.size() ? setWithV : setWithoutV;
+    }
+
+    // TODO: book counts v twice in second case
+    NetworKit::node w;
+    graph->forEdges([&](NetworKit::node a, NetworKit::node b) {
+        if (graph->degree(a) != graph->degree(b)) {
+            if (graph->degree(a) == 5) {
+                v = a;
+                w = b;
+            }
+            else {
+                v = b;
+                w = a;
+            }
+        }        
+    });
```

Comment:
I don't understand this - to be honest, this preserves only the last edge incident to a vertex with deg 5? (properly sorted, of course)

Reply:
Well, wouldn't it be faster to look for the first such edge?

And if so, consider using "for (auto v = graph->nodeRange().begin(); v != graph->nodeRange().end(); ++v) {...}